src/ngraph/pass/algebraic_simplification.cpp

```diff
@@ -419,6 +432,32 @@ static bool simplify_log(shared_ptr<Node> n)
     return false;
 }
 
+// optimizes `gather->shapeof` into `shapeof->gather` for 0D gather indices
+static bool simplify_gather_shapeof(shared_ptr<Node> node)
+{
+    auto shapeof = as_type_ptr<op::v0::ShapeOf>(node);
+    auto gather = as_type_ptr<op::v1::Gather>(shapeof->input_value(0).get_node_shared_ptr());
+    if (!gather)
+    {
+        return false;
+    }
+    auto indices = as_type_ptr<op::Constant>(gather->input_value(1).get_node_shared_ptr());
+    if (!indices || indices->get_shape() != Shape{} ||
+        gather->get_axis() == std::numeric_limits<int64_t>::max())
+    {
+        NGRAPH_DEBUG << gather << " cannot simplify shapeof->gather";
+        return false;
+    }
+    auto new_shapeof = make_shared<op::v0::ShapeOf>(gather->input_value(0).get_node_shared_ptr());
+    auto new_axis = op::Constant::create<int64_t>(element::i64, Shape{}, {0});
+    std::vector<int64_t> vi(gather->get_shape().size() + 1);
+    std::iota(vi.begin(), vi.end(), 0);
+    vi.erase(vi.begin() + gather->get_axis());
```

Comment:
After dynamic shape enablement we should check for negative axis.

Reply:
if input rank is not dynamic, get_axis will adjust negative axis.